src/hotspot/os/windows/os_windows.cpp

```diff
@@ -3982,14 +3982,201 @@ bool   os::win32::_is_windows_server         = false;
+  if (non_null_chars_written == 0) {
+    warning("GetSystemDirectoryW() failed: GetLastError->%ld.", GetLastError());
+    goto free_mem;
+  }
+
+  wcsncat(kernel32_path, L"\\kernel32.dll", filename_chars);
+
+  version_size = GetFileVersionInfoSizeW(kernel32_path, NULL);
+  if (version_size == 0) {
+    warning("GetFileVersionInfoSizeW() failed: GetLastError->%ld.", GetLastError());
+    goto free_mem;
+  }
+
+  version_info = (LPWSTR)os::malloc(version_size, mtInternal);
+  if (version_info == NULL) {
+    warning("os::malloc() failed to allocate %ld bytes for GetFileVersionInfoW buffer", version_size);
+    goto free_mem;
+  }
+
+  if (!GetFileVersionInfoW(kernel32_path, 0, version_size, version_info)) {
+    warning("GetFileVersionInfoW() failed: GetLastError->%ld.", GetLastError());
+    goto free_mem;
+  }
+
+  VS_FIXEDFILEINFO *file_info;
+  if (!VerQueryValueW(version_info, L"\\", (LPVOID*)&file_info, &size)) {
+    warning("VerQueryValueW() failed. Cannot determine Windows version.");
+    goto free_mem;
+  }
+
+  _major_version = HIWORD(file_info->dwProductVersionMS);
+  _minor_version = LOWORD(file_info->dwProductVersionMS);
+  _build_number  = HIWORD(file_info->dwProductVersionLS);
+  _build_minor   = LOWORD(file_info->dwProductVersionLS);
+
+free_mem:
+  os::free(version_info);
+  os::free(kernel32_path);
+}
+
+bool os::win32::is_windows_11_or_greater() {
+  initialize_windows_version();
+
+  // Windows 11 starts at build 22000 (Version 21H2) as per
+  // https://learn.microsoft.com/en-us/windows/release-health/windows11-release-information
+  return (_major_version >= 10 && _build_number >= 22000 && !IsWindowsServer());
+}
+
+bool os::win32::is_windows_server_2022_or_greater() {
+  initialize_windows_version();
+
+  // Windows Server 2022 starts at build 20348.169 as per
+  // https://learn.microsoft.com/en-us/windows/release-health/release-information
+  return (_major_version >= 10 && _build_number >= 20348 && IsWindowsServer());
+}
+
+int os::win32::bit_count(ULONG64 argument) {
+  int bitcount = 0;
+  while (argument != 0) {
+    argument &= (argument-1);
+    bitcount++;
+  }
+  return bitcount;
+}
+
+DWORD os::win32::active_processors_in_job_object() {
+  DWORD processors = 0;
+
+  LPVOID job_object_information = NULL;
+  DWORD job_object_information_length = 0;
+
+  if (!QueryInformationJobObject(NULL, JobObjectGroupInformationEx, NULL, 0, &job_object_information_length)) {
+    DWORD last_error = GetLastError();
+    if (last_error == ERROR_INSUFFICIENT_BUFFER) {
+      DWORD group_count = job_object_information_length / sizeof(GROUP_AFFINITY);
+
+      job_object_information = os::malloc(job_object_information_length, mtInternal);
+      if (job_object_information != NULL) {
+          if (QueryInformationJobObject(NULL, JobObjectGroupInformationEx, job_object_information, job_object_information_length, &job_object_information_length)) {
+            group_count = job_object_information_length / sizeof(GROUP_AFFINITY);
```

Comment:
Will the result of this calculation ever be different from the identical one above?

Reply:
It should be identical. I will convert this into an assertion.